torch/csrc/autograd/engine.cpp

```diff
@@ -944,6 +944,12 @@ void Engine::mark_graph_task_completed(const std::shared_ptr<GraphTask>& graph_t
     graph_task->future_result_->markCompleted(
         std::move(graph_task->captured_vars_));
   } catch (std::exception& e) {
+    // If the callbacks (triggered by markCompleted) throw an exception, just
+    // propagate it further. If we don't do this, setErrorIfNeeded would just
+    // swallow the exception since the future is marked as completed.
+    if (graph_task->future_result_->completed()) {
+      std::rethrow_exception(std::current_exception());
```

Comment:
But what happens if this is running on a different thread? We're going to end up in a bad state here no?

Reply:
If it is running on a different thread, something upstream will catch the exception and LOG it. In case of at::launch(), this would be logged by the c10 ThreadPool. What sort of bad state are you worried about here?

Reply:
What I am wondering is if this could kill our GPU worker threads? Leading to queues without any worker thread working on it.

Reply:
Hmm, I didn't realize the worker threads don't handle exceptions appropriately. I guess we should handle exceptions in those threads since there are other things that could potentially throw apart from this. Although, for this PR I'll remove this line since the changes in `setErrorIfNeeded` would still log the error.

Reply:
If it happens then the error needs to be caught and rethrown in the main thread.

Reply:
future callback always happens in the same thread as the thread marking it complete?

Reply:
Currently it does, but I don't think it is guaranteed.